src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs

```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    public class AssemblyPart : ApplicationPart, IExportTypes
+    {
+        private readonly Assembly _assembly;
+
+        public AssemblyPart(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            _assembly = assembly;
+        }
+
+        public override string Name => _assembly.GetName().Name;
+
+        public IEnumerable<TypeInfo> Types => _assembly.DefinedTypes;
```

Comment:
I chose DefinedTypes for several reasons:
* We were using different APIs on different parts of the codebase.
* DefinedTypes gives you TypeInfo instances which is more aligned with what we use in all our descriptor classes. (An the only thing that contains actual useful stuff, so we can save people from typing a lot of GetTypeInfo())
* All our components (in the future features) already contain logic to ensure that the type meets their specific needs (public, non open genetic, non static, etc)
* Using GetExportedTypes make some customization scenarios a bit harder (like using non public types as controllers) as users would need to write their own part to expose the private types and their feature to select the ones they want.